hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/observer/ObservableImpl.java

```
@@ -0,0 +1,271 @@
+            this.logger = logger;
+            this.executor = getExecutor(hzInstance);
+
+            this.sequence = ringbuffer.headSequence();
+        }
+
+        void next() {
+            if (cancelled) {
+                return;
+            }
+            ringbuffer.readManyAsync(sequence, 1, BATCH_SIZE, null)
+                    .whenCompleteAsync(this::accept, executor);
+        }
+
+        void cancel() {
+            cancelled = true;
+        }
+
+        private void accept(ReadResultSet<Object> resultSet, Throwable throwable) {
+            if (cancelled) {
+                return;
+            }
+            if (throwable == null) {
+                // we process all messages in batch. So we don't release the thread and reschedule ourselves;
+                // but we'll process whatever was received in 1 go.
+                for (Object result : resultSet) {
+                    try {
+                        onNewMessage(result);
+                    } catch (Throwable t) {
+                        logger.warning("Terminating message listener on ring-buffer " + ringbuffer.getName() +
+                                ". Reason: Unhandled exception, message: " + t.getMessage(), t);
+                        cancel();
+                        return;
+                    }
+
+                    sequence++;
+                }
+                next();
+            } else {
+                if (handleInternalException(throwable)) {
+                    next();
+                } else {
+                    cancel();
+                }
+            }
+        }
+
+        private void onNewMessage(Object message) {
+            if (message instanceof Throwable) {
+                observer.onError((Throwable) message);
+            } else if (message instanceof DoneItem) {
+                observer.onComplete();
+            } else {
+                //noinspection unchecked
+                observer.onNext((T) message);
+            }
```

We should log and ignore errors in user's callbacks.